pkg/cmd/experimental/login/smart_merge.go

```diff
@@ -0,0 +1,142 @@
+			requestedClusterNamesToActualClusterNames[requestedKey] = existingName
+			continue
+		}
+
+		uniqueName := getUniqueName(requestedKey, existingClusterNames)
+		requestedClusterNamesToActualClusterNames[requestedKey] = uniqueName
+		ret.Clusters[uniqueName] = needle
+	}
+
+	requestedAuthInfoNamesToActualAuthInfoNames := map[string]string{}
+	existingAuthInfoNames, err := getMapKeys(reflect.ValueOf(haystack.AuthInfos))
+	if err != nil {
+		return nil, err
+	}
+	for requestedKey, needle := range addition.AuthInfos {
+		if existingName := FindExistingAuthInfoName(haystack, needle); len(existingName) > 0 {
+			requestedAuthInfoNamesToActualAuthInfoNames[requestedKey] = existingName
+			continue
+		}
+
+		uniqueName := getUniqueName(requestedKey, existingAuthInfoNames)
+		requestedAuthInfoNamesToActualAuthInfoNames[requestedKey] = uniqueName
+		ret.AuthInfos[uniqueName] = needle
+	}
+
+	requestedContextNamesToActualContextNames := map[string]string{}
+	existingContextNames, err := getMapKeys(reflect.ValueOf(haystack.Contexts))
+	if err != nil {
+		return nil, err
+	}
+	for requestedKey, needle := range addition.Contexts {
+		exists := false
+
+		actualContext := clientcmdapi.NewContext()
+		actualContext.AuthInfo, exists = requestedAuthInfoNamesToActualAuthInfoNames[needle.AuthInfo]
+		if !exists {
+			actualContext.AuthInfo = needle.AuthInfo
+		}
+		actualContext.Cluster, exists = requestedClusterNamesToActualClusterNames[needle.Cluster]
+		if !exists {
+			actualContext.Cluster = needle.Cluster
+		}
+		actualContext.Namespace = needle.Namespace
+		actualContext.Extensions = needle.Extensions
+
+		if existingName := FindExistingContextName(haystack, *actualContext); len(existingName) > 0 {
+			// if this already exists, just move to the next, our job is done
+			requestedContextNamesToActualContextNames[requestedKey] = existingName
+			continue
+		}
+
+		uniqueName := getUniqueName(actualContext.Cluster+"-"+actualContext.AuthInfo, existingContextNames)
+		requestedContextNamesToActualContextNames[requestedKey] = uniqueName
+		ret.Contexts[uniqueName] = *actualContext
+	}
+
+	if len(addition.CurrentContext) > 0 {
+		if newCurrentContext, exists := requestedContextNamesToActualContextNames[addition.CurrentContext]; exists {
+			ret.CurrentContext = newCurrentContext
+		} else {
+			ret.CurrentContext = addition.CurrentContext
+		}
+	}
+
+	return &ret, nil
+}
+
+// FindExistingClusterName finds the nickname for the passed cluster config
+func FindExistingClusterName(haystack clientcmdapi.Config, needle clientcmdapi.Cluster) string {
+	for key, cluster := range haystack.Clusters {
+		if reflect.DeepEqual(cluster, needle) {
+			return key
+		}
+	}
+
+	return ""
+}
+
+// FindExistingAuthInfoName finds the nickname for the passed auth info
+func FindExistingAuthInfoName(haystack clientcmdapi.Config, needle clientcmdapi.AuthInfo) string {
+	for key, authInfo := range haystack.AuthInfos {
+		if reflect.DeepEqual(authInfo, needle) {
```

Comment:
Need to make sure we are using the same equality check rules used in other parts, i.e. when adding users (e.g. nickname forces/ignores letter case, etc). I didn't check.